source/extensions/filters/http/sxg/filter.cc

```diff
@@ -0,0 +1,466 @@
+
+  if (client_accept_sxg_ && shouldEncodeSXG(headers)) {
+    should_encode_sxg_ = true;
+    response_headers_ = &headers;
+    config_->stats().total_should_sign_.inc();
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus Filter::encodeData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(debug, "sxg filter from encodeData end_stream: {}", end_stream);
+
+  if (!should_encode_sxg_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  data_total_ += data.length();
+  if (encoderBufferLimitReached(data_total_)) {
+    should_encode_sxg_ = false;
+    return Http::FilterDataStatus::Continue;
+  }
+
+  encoder_callbacks_->addEncodedData(data, false);
+
+  if (!end_stream) {
+    // We need to know the size of the response in order to generate the SXG, so we wait.
+    return Http::FilterDataStatus::StopIterationAndBuffer;
+  }
+
+  doSxg();
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus Filter::encodeTrailers(Http::ResponseTrailerMap&) {
+  if (should_encode_sxg_) {
+    doSxg();
+  }
+  return Http::FilterTrailersStatus::Continue;
+}
+
+constexpr uint64_t ONE_DAY_IN_SECONDS = 86400L;
+
+void Filter::doSxg() {
+  if (finished_) {
+    return;
+  }
+
+  finished_ = true;
+  encoder_callbacks_->modifyEncodingBuffer([this](Buffer::Instance& enc_buf) {
+    config_->stats().total_signed_attempts_.inc();
+
+    sxg_encoded_response_t encoded = sxg_empty_encoded_response();
+    if (!getEncodedResponse(enc_buf, &encoded)) {
+      sxg_encoded_response_release(&encoded);
+      config_->stats().total_signed_failed_.inc();
+      return;
+    }
+
+    // backdate timestamp by 1 day, to account for clock skew
+    const uint64_t date = getTimestamp() - ONE_DAY_IN_SECONDS;
+
+    sxg_signer_list_t signers = sxg_empty_signer_list();
+    if (!loadSigner(date, &signers)) {
+      sxg_signer_list_release(&signers);
+      sxg_encoded_response_release(&encoded);
+      config_->stats().total_signed_failed_.inc();
+      return;
+    }
+
+    sxg_buffer_t result = sxg_empty_buffer();
+    if (!writeSxg(&signers, url_, &encoded, &result)) {
+      sxg_signer_list_release(&signers);
+      sxg_encoded_response_release(&encoded);
+      sxg_buffer_release(&result);
+      config_->stats().total_signed_failed_.inc();
+      return;
+    }
+
+    sxg_signer_list_release(&signers);
+    sxg_encoded_response_release(&encoded);
+
+    // Make sure that the resulting SXG isn't too big before adding it to the encoding buffer
+    if (encoderBufferLimitReached(result.size + 100)) {
+      config_->stats().total_signed_failed_.inc();
+      sxg_buffer_release(&result);
+      return;
+    }
+
+    const absl::string_view output(reinterpret_cast<char*>(result.data), result.size);
+
+    enc_buf.drain(enc_buf.length());
+    enc_buf.add(output); // TODO(rgs): can we move this and avoid the copy?
```

Comment:
I'm not familiar with Envoy but it looks like `Buffer::Instance::move()` takes another `Buffer::Instance`. (I don't see an easy way to construct that from a `string_view`, but I may have missed it.)

Perhaps [`addBufferFragment()`](https://github.com/envoyproxy/envoy/blob/6834081f835d609606e77ef84d734e30fc403d58/envoy/buffer/buffer.h#L150-L155) could be used? It looks like `enc_buf` needs to outlive this lambda, so I think you'll need to allocate the `sxg_buffer_t` on the heap and then construct a [`BufferFragmentImpl`](https://github.com/envoyproxy/envoy/blob/0fae6970ddaf93f024908ba304bbd2b34e997a51/source/common/buffer/buffer_impl.h#L598) whose `releasor` calls something like `sxg_buffer_release(result); delete result;`. Does that sound right?

I think whether you can can replace `sxg_buffer_release` with a call to `free(result.data)` depends on whether `OPENSSL_free` has been overridden ([BoringSSL](https://github.com/google/boringssl/blob/c1571feb5faf5cce844354c63d0f3e842464bea3/crypto/mem.c#L155-L166); [OpenSSL](https://github.com/openssl/openssl/blob/1c0eede9827b0962f1d752fa4ab5d436fa039da4/crypto/mem.c#L247-L250)).

Reply:
`addBufferFragment(...)` is working for me locally. `result.data` is already allocated on the heap, right? so if it's safe to call `OPENSSL_free` on `result.data` directly (rather than calling `sxg_buffer_release`), then we're good. if not, we can either allocate the `sxg_buffer_t` on the heap or add it to the filter class.

Reply:
Ah yeah, you're right; we don't need to move the `sxg_buffer_t` to the heap. I guess either option works:
 1. Capture the `sxg_buffer_t` by value into the `releasor` lambda, and use the standard `sxg_buffer_release`.
 2. Capture the `uint8_t*` by value and use `OPENSSL_free`.

(1) seems slightly nicer since it adheres to the higher-level [documented contract](https://github.com/google/libsxg/blob/672b807e9a1a0032c73f1eeb7c9761d8cfd013d0/include/libsxg/sxg_buffer.h#L39-L40) in sxg_buffer.h, so less likely to break with a dependency update. But if that's suboptimal for some other reason (performance?) then (2) works for me, too. In that case, maybe send a PR to update the comment in sxg_buffer.h so it becomes a part of the contract.